runtime/lua/vim/func.lua

```diff
@@ -3,40 +3,11 @@ local M = {}
 -- TODO(lewis6991): Private for now until:
 -- - There are other places in the codebase that could benefit from this
 --   (e.g. LSP), but might require other changes to accommodate.
--- - Invalidation of the cache needs to be controllable. Using weak tables
---   is an acceptable invalidation policy, but it shouldn't be the only
---   one.
 -- - I don't think the story around `hash` is completely thought out. We
 --   may be able to have a good default hash by hashing each argument,
 --   so basically a better 'concat'.
 -- - Need to support multi level caches. Can be done by allow `hash` to
 --   return multiple values.
---
---- Memoizes a function {fn} using {hash} to hash the arguments.
----
---- Internally uses a |lua-weaktable| to cache the results of {fn} meaning the
---- cache will be invalidated whenever Lua does garbage collection.
----
---- The memoized function returns shared references so be wary about
---- mutating return values.
----
---- @generic F: function
---- @param hash integer|string|function Hash function to create a hash to use as a key to
----     store results. Possible values:
----     - When integer, refers to the index of an argument of {fn} to hash.
----     This argument can have any type.
----     - When function, is evaluated using the same arguments passed to {fn}.
----     - When `concat`, the hash is determined by string concatenating all the
----     arguments passed to {fn}.
----     - When `concat-n`, the hash is determined by string concatenating the
----     first n arguments passed to {fn}.
----
---- @param fn F Function to memoize.
---- @param strong? boolean Do not use a weak table
---- @return F # Memoized version of {fn}
---- @nodoc
-function M._memoize(hash, fn, strong)
-  return require('vim.func._memoize')(hash, fn, strong)
-end
+M._memoize = require('vim.func._memoize')
```

Comment:
The point of wrapping this in a function is so the `require` is only done when the function is called. I'd prefer if we keep this as it was for now.

Reply:
maybe add a `-- lazy-load` comment so this is clear

Reply:
why not move the memoize impl to `runtime/lua/vim/func.lua` ?

Reply:
I put it in its own file (`runtime/lua/vim/func/_memoize.lua`) for better encapsulation as `_memoize` needs at least 3 niche helper functions, and probably more in the future when the other TODO's are implemented.

Reply:
I'm fine to change this back, but I had initially changed it also to reduce `require` usage; I may be misunderstanding but wouldn't the previous functionality always call `require` on each function call, whereas this version only calls `require` the one time initially (since it is not wrapped in a function, and thus just gets evaluated as a value the one time)? My debugging seems to confirm this as well, using

```lua
my_count = 0

local require = function(module)
  my_count = my_count + 1
  vim.print(my_count)
  return original_require(module)
end
```

I may be misunderstanding something, please let me know

Reply:
>  but wouldn't the previous functionality always call require on each function call

`require` is self caching so this isn't a problem. On the first call `required` places the module in `package.loaded[<module_name>]`, and fetches from that on subsequent calls.

Reply:
If you want to optimize away the table indexes, which I don't recommend, you could do.

```lua
function M._memoize(hash, fn, strong)
  M._memoize = require('vim.func._memoize')
  return M._memoize(hash, fn, strong)
end
```

But this is an extreme micro-op IMO.

Reply:
Ah, I understand the purpose of the original functionality then, I misinterpreted what you meant initially (deferred require > less requires). Thanks